### PR TITLE
Async rst async queue

### PR DESCRIPTION
**Pull request by mwachs5**
@terpstra @yunsup 

In general, it's a good idea to have asynchronous reset on the crossing FIFOs, since the blocks may come out of reset at different times, clocks may not have started yet, etc.

This PR achieves that and blows away the non-async-reset option.

Are there any other low-level crossing primitives that need this treatment?

### CODE REVIEW

File: src/main/scala/junctions/asyncqueue.scala [hunk 1]
```diff
@@ -66,9 +75,13 @@ class AsyncQueueSink[T <: Data](gen: T, depth: Int, sync: Int, clockIn: Clock, r
   // On an FPGA, only one input changes at a time => mem updates don't cause glitches
   // The register only latches when the selected valued is not being written
   val index = if (depth == 1) UInt(0) else ridx(bits-1, 0) ^ (ridx(bits, bits) << (bits-1))
-  io.deq.bits  := RegEnable(io.mem(index), valid && !reset)
-  io.deq.valid := RegNext(valid, Bool(false))
-  io.ridx := RegNext(ridx, UInt(0))
+  io.deq.bits  := RegEnable(io.mem(index), valid && !reset) // This does NOT need to be reset.
```
Comment (mwachs5) [inline src/main/scala/junctions/asyncqueue.scala]: Actually @terpstra why is *this* reset included in the enable logic?

Comment (mwachs5) [inline src/main/scala/junctions/asyncqueue.scala]: @terpstra I removed this reset. Let me know if there was something I was missing there.

**Comment by terpstra** [comment]
How urgently do you need this merged? I'd like to go over it very carefully on Monday.

File: src/main/scala/junctions/asyncqueue.scala [hunk 2]
```diff
@@ -2,22 +2,28 @@
...
 import Chisel._
+import uncore.util.{AsyncResetRegVec, AsyncResetReg}
 
 object GrayCounter {
   def apply(bits: Int, increment: Bool = Bool(true)): UInt = {
-    val binary = RegInit(UInt(0, width = bits))
-    val incremented = binary + increment.asUInt()
-    binary := incremented
+    val incremented = Wire(UInt(width=bits))
+    val binary = AsyncResetReg(incremented, 0, Bool(true))
```
Comment (terpstra) [inline src/main/scala/junctions/asyncqueue.scala]: Wouldn't
  val binary = AsyncResetReg(incremented)
be enough?

File: src/main/scala/junctions/asyncqueue.scala [hunk 3]
```diff
@@ -33,15 +39,19 @@ class AsyncQueueSource[T <: Data](gen: T, depth: Int, sync: Int, clockIn: Clock,
...
   val index = if (depth == 1) UInt(0) else io.widx(bits-1, 0) ^ (io.widx(bits, bits) << (bits-1))
   when (io.enq.fire() && !reset) { mem(index) := io.enq.bits }
-  io.enq.ready := RegNext(ready, Bool(false))
-  io.widx := RegNext(widx, UInt(0))
+  val ready_reg = AsyncResetReg(ready, 0)
+  io.enq.ready := ready_reg
+
+  val widx_reg = AsyncResetReg(widx, 0)
+  io.widx := widx_reg
+
```
Comment (terpstra) [inline src/main/scala/junctions/asyncqueue.scala]: Why did you add intermediate variable names? :-(

File: src/main/scala/junctions/asyncqueue.scala [hunk 4]
```diff
@@ -66,9 +76,16 @@ class AsyncQueueSink[T <: Data](gen: T, depth: Int, sync: Int, clockIn: Clock, r
   // On an FPGA, only one input changes at a time => mem updates don't cause glitches
   // The register only latches when the selected valued is not being written
   val index = if (depth == 1) UInt(0) else ridx(bits-1, 0) ^ (ridx(bits, bits) << (bits-1))
-  io.deq.bits  := RegEnable(io.mem(index), valid && !reset)
-  io.deq.valid := RegNext(valid, Bool(false))
-  io.ridx := RegNext(ridx, UInt(0))
+  // This register does not NEED to be reset, as its contents will not
+  // be considered unless deq_reg is true.
+  io.deq.bits  := RegEnable(io.mem(index), valid)
```
Comment (terpstra) [inline src/main/scala/junctions/asyncqueue.scala]: The !reset was there as a work-around for the fact that I did not have async reset. I didn't want it latching garbage. You were right to remove it.

File: src/main/scala/junctions/asyncqueue.scala [hunk 5]
```diff
@@ -66,9 +76,16 @@ class AsyncQueueSink[T <: Data](gen: T, depth: Int, sync: Int, clockIn: Clock, r
...
+  // This register does not NEED to be reset, as its contents will not
+  // be considered unless deq_reg is true.
+  io.deq.bits  := RegEnable(io.mem(index), valid) 
+    
+
+  val deq_reg = AsyncResetReg(valid, 0)
+  io.deq.valid := deq_reg
+
+  val ridx_reg = AsyncResetReg(ridx, 0)
+  io.ridx := ridx_reg
```
Comment (terpstra) [inline src/main/scala/junctions/asyncqueue.scala]: You just like making my pretty code ugly, don't you?

File: src/main/scala/junctions/asyncqueue.scala [hunk 6]
```diff
@@ -2,22 +2,28 @@
...
-    syncv(0)
+    val syncv = List.fill(sync)(Module (new AsyncResetRegVec(w = in.getWidth, 0)))
+    syncv.last.io.d := in
+    syncv.last.io.en := Bool(true)
+      (syncv.init zip syncv.tail).foreach { case (sink, source) => {
+        sink.io.d := source.io.q
+        sink.io.en := Bool(true)
+      }
+      }
+    syncv(0).io.d
```
Comment (terpstra) [inline src/main/scala/junctions/asyncqueue.scala]: This is rather hideous. Oh well...

Comment (mwachs5) [inline src/main/scala/junctions/asyncqueue.scala]: Yeah. I tried to think of some Scala fanciness, then decided sometimes you should just write what you mean.

File: src/main/scala/junctions/asyncqueue.scala (hunk 3 above)
Comment (mwachs5) [inline src/main/scala/junctions/asyncqueue.scala]: ah, this was the remnant of my having actual module instantiations before I realized there were nice objects now. I'll tidy up.

File: src/main/scala/junctions/asyncqueue.scala (hunk 2 above)
Comment (mwachs5) [inline src/main/scala/junctions/asyncqueue.scala]: I like to be clear about the reset value. But you're right about the Bool(true). I removed that.

File: src/main/scala/junctions/asyncqueue.scala (hunk 3 above)
Comment (mwachs5) [inline src/main/scala/junctions/asyncqueue.scala]: tidied.
